counsel.el: Add directory change support to counsel-git

brownts opened the pull request:
Fixes #2740.

Adds a keymap for `counsel-git` allowing `counsel-cd` to be invoked.  Additionally, `counsel-git` has been modified to allow an optional initial directory to be provided (as a result of invoking `counsel-cd`).

basil-conto commented on counsel.el:
```diff
@@ -1252,6 +1252,11 @@ selected face."
 
 ;;* Git
 ;;** `counsel-git'
+(defvar counsel-git-map
+  (let ((map (make-sparse-keymap)))
+    (define-key map (kbd "C-x C-d") 'counsel-cd)
+    map))
```
Nit: please always include a docstring.

basil-conto commented on counsel.el:
```diff
@@ -1280,14 +1285,18 @@ Like `locate-dominating-file', but DIR defaults to
   (counsel-require-program counsel-git-cmd)
-  (let ((default-directory (counsel-locate-git-root)))
+  (let ((enable-recursive-minibuffers t)
+        (default-directory (or initial-directory
+                               (counsel-locate-git-root))))
```
This doesn't seem to work with `counsel-git-action`:

1. `make plain`
2. `M-x counsel-git RET`
3. `C-x C-d doc RET RET`

This results in `doc/doc/Changelog.org` being visited, rather than the expected `doc/Changelog.org`.

basil-conto reviewed (changes requested):
Thanks, see some comments below.

brownts commented:
Thanks @basil-conto, I have now added a docstring to `counsel-git-map`.  Additionally, nice catch on the second issue.  I only tested at the git root level, so missed the issue when you move into a sub-directory.  I have now addressed that by adjusting the `counsel-git-cmd` to not generate the complete path from the git root (via `--full-name`), but only relative to the directory where the search was started.

basil-conto reviewed (approved):
Thanks, LGTM.

I'm not sure whether I can merge this without a copyright assignment to the FSF, though, given your cumulative changes to this project. Do you have a CA file already? If not, would you be willing to? If you're unfamiliar with this, see [`(info "(emacs) Copyright Assignment")`](https://www.gnu.org/software/emacs/manual/html_node/emacs/Copyright-Assignment.html). Thanks.

brownts commented:
I don't use this package anymore, so no longer have a need for this.
